Document testing infrastructure (#2279)
input-output-hk/ouroboros-network: ouroboros-consensus/docs/Testing.md

create a directory, open a file, write some bytes, etc. The model is a very
simple one: we simply model a file system as a tree of `ByteString`s (the raw
bytes in the files). The goal of the model, of course, is not to provide one that
mirrors the low-level details of the real implementation, but rather one that
abstracts over such details and provides a _specification_.

The file system abstraction however is a bit of an unusual test in that this is
really "reverse model testing": the tests compare the model to the real file
system, but we are of course not developing a file system. Instead, the tests
serve to make sure that we got the model right, which we can then use
in the tests of the rest of the consensus layer.

**Stats.** The implementation, including the OS specific parts, is 1200 loc.
The tests are 1500 loc.

### The Immutable DB (`Test.Ouroboros.Storage.ImmutableDB`)

The immutable DB bundles a (configurable) number of blocks into "chunk files".
By design, chunk files are literally just the raw blocks, one after the other,
so that we can efficiently support binary streaming of blocks.

Every chunk file is accompanied by two indices: a _primary_ index that for

> each slot in the chunk file provides an index into a _secondary_ index, which

Reviewer: ```suggestion
each slot in the chunk file provides an offset into a _secondary_ index, which
```
Double use of "index" might be confusing.

simple: the volatile DB is modelled simply as a set of blocks, and the
immutable DB is modelled simply as a list of blocks (i.e., a chain).

Nonetheless, the implementation of the operations on that model is subtle.
In particular, the chain DB is responsible for chain selection, and so the
model must too. So we have a few properties checking some aspects of the model;
in particular, we verify that no matter in which order we add blocks to the
chain DB, we always pick the most preferred chain.

#### Model based testing (`Test.Ouroboros.Storage.ChainDB.StateMachine`)

These are the main tests for the chain DB. Commands include

* Add a block
* Add a block with a `SlotNo` that is ahead of the wall-clock.
* Get the current chain and/or ledger state
* Create a new iterator and use it to stream blocks
* Create a new reader and use it to follow the chain
* (Limited) disk corruption (the chain DB relies on the immutable DB and
  volatile DB for the storage proper and _they_ have extensive disk corruption
  tests, so we don't need to repeat that here).

The model was described in the previous section.

Note that it is important to tests blocks with a `SlotNo` ahead of the

> wall-click separately, because the Ouroboros protocol says such blocks should

Reviewer: ```suggestion
wallclock separately, because the Ouroboros protocol says such blocks should
```

the wallclock to a `SlotNo`, although we _can_ always translate the `SlotNo`
at the tip of the chain to a `UTCTime`.

## Miscellanous tests (`test-consensus` test suite)

This test suite contains tests for a number of components of the rest of the
consensus layer.

### Resource registry (`Test.Consensus.ResourceRegistry`)

The resource registry is a component throughout the consensus layer that helps
us keep track of resources and makes sure that all resources that we allocate
are eventually also deallocated.

The tests for the registry are model based. The model records which resources
we expect to be alive and which we expect to have been deallocated. The only
resources we are modelling here are threads; the commands we then execute are

* Fork a thread from some other thread
* Terminate a thread
* Have a thread crash
* Collect all live threads

We then verify that the resource registry behaves like the model, cleaning
up resources as threads terminate or crash.

**Stats.** The implementation is 1200 loc, the tests are 600 loc.

### Locking (`Test.Consensus.Util.MonadSTM.RAWLock`)

The volatile DB uses an abstraction we call a `RAWLock`, a lock that allows

> the following combinations of reader, appender and writers:

Reviewer: ```suggestion
the following combinations of readers, appender and writer:
```

```
         │ Reader │ Appender │ Writer │
─────────┼────────┼──────────┼────────┤
Reader   │   V    │     V    │    X   │
Appender │░░░░░░░░│     X    │    X   │
Writer   │░░░░░░░░│░░░░░░░░░░│    X   │
```

It improves concurrent access. In the test we generate lots of threads, some

> readers, some appenders, some writers, each of which concurrently accessing

Reviewer: ```suggestion
readers, some appenders, some writers, each concurrently accessing
```

kinds of queries that can be asked) of course depends on the ledger. The tests
use a mock ledger for this purpose.

**Stats.** The implementation is 75 loc, the tests are 250 loc.

### Versioned serialization (`Test.Consensus.Util.Versioned`)

Some of our serialization code supports some limited migration capability. This
module contains a few unit tests that check that the migration infrastructure we
provide works as expected. There is not much here.

**Stats.** The implementation is 120 loc, the tests are 110 loc.

### DB marker and DB lock (`Test.Consensus.Node`)

When the consensus layer is integrated into the main node, it provides two
safe guards to avoid data loss and/or corruption:

* When the database is opened, it locks the database so that there can be no
  other processes trying to access it at the same time.
* When we create a database directory, we place a "magic marker" in that
  directory. This allows us to distinguish the database directory from other
  directories, and avoids that we would try to "truncate" a "chain" in a
  directory which doesn't contain a DB at all (due to a misconfiguration),
  thereby potentially deleting a user's files.

> This modules contains a bunch of unit tests to make sure that these locks and

Reviewer: ```suggestion
This module contains a bunch of unit tests to make sure that these locks and
```

One of the responsibilities of the HFC is to offer time conversions (slot to
epoch, wall clock to slot, etc.) across era transitions (which might change
parameters such as the slot length and the epoch size). It does this by
constructing a "summary" of the current state of the chain, basically recording
what the various slot lengths and epoch sizes have been so far, and how far
ahead we can look (the so-called "safe zone": if the transition to the next era
is not yet known, there must exist a limited period after the ledger tip in
which we can still time conversions).

**Stats.** The HFC history implementation (not the combinator) is 1300 loc;
the tests are also 1300 loc.

#### Summary (`Test.Consensus.HardFork.Summary`)

This module verifies the property that _no matter how the summary is constructed_,
as long as it satisfies its invariants, we should have roundtrip properties:

* Converting time to a slot and then back to time should be an identity
  (modulo the time spent in that slot)
* Converting a slot to time and then back should be an identity
* Converting slot to an epoch and then back to a slot should be an identity
  (modulo the time spent in that epoch)

> * Converting an epoch to a slot and then back should be an identity.

Reviewer: ```suggestion
* Converting an epoch to a slot and then back should be an identity
```
Or is the period intentional because it's the last bullet?

#### History (`Test.Consensus.HardFork.History`)

This is the more interesting test of the hard fork history. We construct a mock
chain, consisting of events (events are roughly, but not quite, "blocks"). For
every event we record its slot number, epoch number, wall clock, etc. Since we
are constructing this chain as a whole, from genesis to its tip, constructing
these events is trivial. We then split this chain in half, and construct a

> 'Summary' from the first half. We then use that summary to do conversions for

Reviewer: Backticks or double quotes?

but is used to do conversions across the entire chain, we verify that
predictions about the "future" also work as correctly (including that the
conversions say "outside range" if and only if the model expects them to be).

#### `Test.Consensus.Protocol.PBFT`

The PBFT consensus protocol needs to verify that within any given window, there
is no more than a certain threshold of slots signed by a single key. This sounds
easy enough, but it's complicated by two factors: first, it must support
rollback, and second, it must deal with EBBs, a special kind of block on
the Byron chain that shares a `SlotNo` and a `BlockNo` with non-EBB blocks,
leading to all kinds of trouble.

The tests here are property tests that validate that we can always rollback
to the state we're interested in, within the contraints set by `k` (the
security parameter, indicating how far we can rollback), whether or not
EBBs are present.

**Stats.** The PBFT state implementation (not the entire protocol) is 750 loc;
the tests are 700 loc.

## Consensus tests

The main consensus tests are the consensus layer's sophisticated tests. They are
"system level" test, in which we set up a mock network of nodes, where we can

> simulate things like nodes joining late, network message delay, etc. etc. We

Reviewer: ```suggestion
simulate things like nodes joining late, network message delay, etc. We
```

then have these nodes run the full protocol, exchanging blocks, doing chain
selection, the whole shebang, and then verify that the nodes can reach
consensus. We have some generic infrastructure for doing all of this, and then
have specific tests for each of the protocols/ledgers we support.

In a way, these are the most important tests we have, as they are testing
properties of the system as a whole. Of course, that also means that if
something goes wrong, debugging these tests can be difficult, and it would be

> better to have such  problems caught by the other, more fine-grained, tests.

Reviewer: ```suggestion
better to have such problems caught by the other, more fine-grained, tests.
```

We run these tests for

* A mock ledger (containing bare bones UTxO style transactions) using a variety
  of consensus protocols: BFT, PBFT, Praos, and a version of Praos where we
  override the leader schedule.
* Byron/PBFT
* DualByron: This runs the Byron ledger and the Byron specification in lockstep,
  verifying that they agree at every point.

> * Shelley/PTraos

Reviewer: ```suggestion
* Shelley/TPraos
```
Shelley would never betray us!

Author: :D

> * The hard fork combinator transitioning from a mock ledger A and a mock ledger B;

Reviewer: ```suggestion
* The hard fork combinator transitioning from a mock ledger A to a mock ledger B;
```
